File: tests/analyze_tx.rs
//! 分析 transaction.data 中的差异记录
//!
//! 从 tests/transaction.data 文件中读取每笔交易，
//! 解析 attachment_bytes 的二进制结构，与当前 Rust 实现对比

use std::fs::File;
use std::io::{BufRead, BufReader};

fn main() {
    let file = File::open("tests/transaction.data").expect("Failed to open transaction.data");
    let reader = BufReader::new(file);
    
    println!("=== Transaction Data Analysis ===\n");
    println!("{:<6} {:<22} {:<4} {:<4} {:<10} {:<70} {:<8} ATTACHMENT_BYTES_LEN | PARSED_STRUCTURE",
        "DB_ID", "ID", "TYPE", "SUBTYPE", "VERSION", "FULL_HASH", "ATT_BYTES_HEX");
    println!("{}", "-".repeat(200));
    
    for (line_num, line) in reader.lines().enumerate() {
        if line_num == 0 { continue; } // skip header
        
        let fields: Vec<&str> = line.split('\t').collect();
        if fields.len() < 25 { continue; }
        
        let db_id = fields[0].trim();
        let id = fields[1].trim();
        let tx_type = fields[11].trim();
        let subtype = fields[12].trim();
        let version = fields[19].trim();
        let full_hash = fields[6].trim();
        let att_bytes_hex = fields[17].trim();
        
        let att_bytes = match hex::decode(att_bytes_hex) {
            Ok(b) => b,
            Err(_) => continue,
        };
        
        // Parse attachment_bytes structure
        let structure = parse_attachment_structure(tx_type.parse().unwrap_or(0), 
                                                   subtype.parse().unwrap_or(0),
                                                   version.parse().unwrap_or(0),
                                                   &att_bytes);
        
        println!("{:<6} {:<22} {:<4} {:<4} {:<10} {:<70} {:<8} {:<4} | {}",
            db_id, id, tx_type, subtype, version, full_hash,
            &att_bytes_hex[..att_bytes_hex.len().min(40)],
            att_bytes.len(),
            structure);
    }
}

fn parse_attachment_structure(tx_type: u8, subtype: u8, version: u8, bytes: &[u8]) -> String {
    if bytes.is_empty() { return "(empty)".to_string(); }
    
    let mut parts = Vec::new();
    let mut offset = 0;
    
    // First byte is usually attachment version (if version > 0)
    if bytes.len() >= 1 && version > 0 {
        parts.push(format!("att_ver={}", bytes[0]));
        offset = 1;
    }
    
    match (tx_type, subtype) {
        (1, 9) => {
            // PhasingVoteCasting
            if offset < bytes.len() {
                let count = bytes[offset];
                parts.push(format!("count={}", count));
                offset += 1;
                for i in 0..count.min(3) as usize {
                    if offset + 32 <= bytes.len() {
                        parts.push(format!("hash[{}]={}", i, hex::encode(&bytes[offset..offset+32])));
                        offset += 32;
                    }
                }
                if offset < bytes.len() {
                    parts.push(format!("trailing={}", hex::encode(&bytes[offset..])));
                }
            }
        }
        (1, 0) => {
            // ArbitraryMessage - may have Message appendix
            if offset + 4 <= bytes.len() {
                let len_field = i32::from_le_bytes([bytes[offset], bytes[offset+1], bytes[offset+2], bytes[offset+3]]);
                let is_text = len_field < 0;
                let msg_len = (len_field as u32) & 0x7FFFFFFF;
                parts.push(format!("msg_len={} isText={}", msg_len, is_text));
                offset += 4;
                if offset + msg_len as usize <= bytes.len() {
                    let msg_data = &bytes[offset..offset + msg_len as usize];
                    let preview = String::from_utf8_lossy(&msg_data[..msg_data.len().min(16)]);
                    parts.push(format!("msg=\"{}{}\"", preview, if msg_data.len() > 16 { "..." } else { "" }));
                    offset += msg_len as usize;
                }
            }
            if offset < bytes.len() {
                parts.push(format!("extra={}", hex::encode(&bytes[offset..])));
            }
        }
        (12, 0) => {
            // AliasAssignment
            if offset < bytes.len() {
                let name_len = bytes[offset] as usize;
                parts.push(format!("name_len={}", name_len));
                offset += 1;
                if offset + name_len <= bytes.len() {
                    let name = String::from_utf8_lossy(&bytes[offset..offset+name_len]);
                    parts.push(format!("alias=\"{}\"", name));
                    offset += name_len;
                }
            }
            // priceNQT (8 bytes)
            if offset + 8 <= bytes.len() {
                let price = u64::from_le_bytes([
                    bytes[offset], bytes[offset+1], bytes[offset+2], bytes[offset+3],
                    bytes[offset+4], bytes[offset+5], bytes[offset+6], bytes[offset+7],
                ]);
                parts.push(format!("price={}", price));
                offset += 8;
            }
            // May have Message appendix following
            if offset + 4 <= bytes.len() {
                let len_field = i32::from_le_bytes([bytes[offset], bytes[offset+1], bytes[offset+2], bytes[offset+3]]);
                let msg_len = (len_field as u32) & 0x7FFFFFFF;
                if msg_len > 0 && msg_len <= 1000 && offset + 4 + msg_len as usize <= bytes.len() {
                    parts.push(format!("appended_msg_len={}", msg_len));
                }
            }
            if offset < bytes.len() {
                parts.push(format!("extra={}", hex::encode(&bytes[offset..])));
            }
        }
        (2, 10) => {
            // PublicKeyAnnouncement
            if offset + 32 <= bytes.len() {
                let pk = hex::encode(&bytes[offset..offset+32]);
                parts.push(format!("pubkey={}", &pk[..16]));
                offset += 32;
            }
            if offset < bytes.len() {
                parts.push(format!("extra={}", hex::encode(&bytes[offset..])));
            }
        }
        (5, 0) => {
            // AssetTransfer or similar
            if offset + 8 <= bytes.len() {
                let asset_id = u64::from_le_bytes([
                    bytes[offset], bytes[offset+1], bytes[offset+2], bytes[offset+3],
                    bytes[offset+4], bytes[offset+5], bytes[offset+6], bytes[offset+7],
                ]);
                parts.push(format!("asset_id={}", asset_id));
                offset += 8;
            }
            if offset + 8 <= bytes.len() {
                let quantity = u64::from_le_bytes([
                    bytes[offset], bytes[offset+1], bytes[offset+2], bytes[offset+3],
                    bytes[offset+4], bytes[offset+5], bytes[offset+6], bytes[offset+7],
                ]);
                parts.push(format!("quantity={}", quantity));
                offset += 8;
            }
            if offset < bytes.len() {
                parts.push(format!("extra={}", hex::encode(&bytes[offset..])));
            }
        }
        _ => {
            // Generic: show first few bytes and total length
            if offset < bytes.len() {
                let remaining = &bytes[offset..];
                parts.push(format!("data={}B:{}", remaining.len(), hex::encode(&remaining[..remaining.len().min(20)])));
            }
        }
    }
    
    parts.join(" | ")
}

File: tests/analyze_tx.py
#!/usr/bin/env python3
"""分析 transaction.data - 正确解析 attachment_bytes 结构（修正字段索引）"""
import struct

with open('tests/transaction.data') as f:
    lines = f.readlines()

print(f'Total records: {len(lines)-1}')
print()
print("=" * 200)
print(f'{"#":>3} {"T:S":>5} {"VER":>4} {"PHASED":>6} {"PRUN_MSG":>8} {"PRUN_ATT":>8} | {"ATT_LEN":>6} | DECODED_STRUCTURE')
print("=" * 200)

for i, line in enumerate(lines[1:], 1):
    fields = [f.strip() for f in line.strip().split('|')]
    if len(fields) < 29:
        continue
    
    tx_type = int(fields[11])
    subtype = int(fields[12])
    version = int(fields[19])
    full_hash = fields[6][:24]
    att_hex = fields[18]  # ATTACHMENT_BYTES is at index 18!
    phased = fields[17]
    has_prunable_msg = fields[23]
    has_prunable_att = fields[24]
    
    try:
        if not att_hex or att_hex.upper() == 'FALSE' or att_hex.upper() == 'NULL':
            print(f'#{i:>3} {tx_type}:{subtype:<2} {version:>4} {phased:>6} {has_prunable_msg:>8} {has_prunable_att:>8} | {"EMPTY":>6} | *** NO BINARY DATA ***')
            continue
            
        att_bytes = bytes.fromhex(att_hex)
        off = 0
        parts = []
        
        if len(att_bytes) > 0 and version == 1:
            parts.append(f'ver={att_bytes[0]:02x}')
            off = 1
        
        if tx_type == 1 and subtype == 9:  # PhasingVoteCasting
            if off < len(att_bytes):
                cnt = att_bytes[off]; off += 1
                parts.append(f'count={cnt}')
                for c in range(min(cnt, 3)):
                    if off + 32 <= len(att_bytes):
                        h = att_bytes[off:off+32].hex()
                        parts.append(f'hash[{c}]={h[:20]}..')
                        off += 32
                if off < len(att_bytes):
                    trailing = att_bytes[off:].hex()
                    tlen = len(att_bytes) - off
                    parts.append(f'TRAIL[{tlen}B]={trailing}')
        
        elif tx_type == 1 and subtype == 0:  # ArbitraryMessage (may have Message appendix)
            if off + 4 <= len(att_bytes):
                lf = struct.unpack_from('<i', att_bytes, off)[0]
                is_text = lf < 0
                msg_len = lf & 0x7fffffff
                parts.append(f'msg_len={msg_len} txt={is_text}')
                off += 4
                if off + msg_len <= len(att_bytes):
                    raw = att_bytes[off:off+msg_len]
                    try:
                        preview = raw.decode('utf8', errors='replace')[:25]
                        parts.append(f'msg="{preview}"')
                    except:
                        parts.append(f'msg_hex={raw.hex()[:30]}')
                    off += msg_len
            if off < len(att_bytes):
                parts.append(f'extra={att_bytes[off:].hex()[:30]}')
        
        elif tx_type == 12 and subtype == 0:  # AliasAssignment
            if off < len(att_bytes):
                nl = att_bytes[off]; off += 1; parts.append(f'name_len={nl}')
                if off + nl <= len(att_bytes):
                    name = att_bytes[off:off+nl].decode('utf8', errors='replace')
                    parts.append(f'alias="{name}"'); off += nl
            if off + 8 <= len(att_bytes):
                price = struct.unpack_from('<Q', att_bytes, off)[0]; off += 8
                parts.append(f'price={price}')
            if off < len(att_bytes):
                parts.append(f'extra={att_bytes[off:].hex()[:30]}')
        
        elif tx_type == 2 and subtype == 10:  # PublicKeyAnnouncement
            if off + 32 <= len(att_bytes):
                pk = att_bytes[off:off+32].hex(); off += 32
                parts.append(f'pk={pk[:28]}..')
            if off < len(att_bytes):
                parts.append(f'extra={att_bytes[off:].hex()[:30]}')
        
        elif tx_type == 5 and subtype == 0:  # AssetTransfer
            if off + 16 <= len(att_bytes):
                aid = struct.unpack_from('<Q', att_bytes, off)[0]; off += 8
                qty = struct.unpack_from('<Q', att_bytes, off)[0]; off += 8
                parts.append(f'asset={aid} qty={qty}')
            if off < len(att_bytes):
                parts.append(f'extra={att_bytes[off:].hex()[:30]}')
        
        elif tx_type == 2 and subtype == 0:  # AssetIssuance
            if off < len(att_bytes):
                nl = att_bytes[off]; off += 1; parts.append(f'name_len={nl}')
                if off + nl <= len(att_bytes):
                    name = att_bytes[off:off+nl].decode('utf8', errors='replace')
                    parts.append(f'name="{name}"'); off += nl
            if off + 2 <= len(att_bytes):
                dl = struct.unpack_from('<H', att_bytes, off)[0]; off += 2; parts.append(f'desc_len={dl}')
                if off + dl <= len(att_bytes):
                    desc = att_bytes[off:off+dl].decode('utf8', errors='replace')[:15]
                    parts.append(f'desc="{desc}"'); off += dl
            if off + 8 <= len(att_bytes):
                qty = struct.unpack_from('<Q', att_bytes, off)[0]; off += 8; parts.append(f'qty={qty}')
            if off < len(att_bytes):
                dec = att_bytes[off]; off += 1; parts.append(f'decimals={dec}')
            if off < len(att_bytes):
                parts.append(f'extra={att_bytes[off:].hex()[:30]}')
        
        elif tx_type == 6 and subtype == 0:  # Some type with prunable attachment
            if off < len(att_bytes):
                remaining = att_bytes[off:]
                parts.append(f'data={len(remaining)}B:{remaining.hex()[:40]}')
        
        else:
            if off < len(att_bytes):
                parts.append(f'raw={att_bytes[off:].hex()[:45]}')
        
        print(f'#{i:>3} {tx_type}:{subtype:<2} {version:>4} {phased:>6} {has_prunable_msg:>8} {has_prunable_att:>8} | {len(att_bytes):>6}B | {" | ".join(parts)}')
        
    except Exception as e:
        print(f'#{i:>3} {tx_type}:{subtype:<2} {version:>4} | ERR: {e}')

print("\n" + "=" * 200)
